src/storage/log/mod.rs

```diff
@@ -0,0 +1,2388 @@
+        self
+    }
+
+    pub fn with_dir_mode(mut self, dir_mode: u32) -> Self {
+        self.dir_mode = Some(dir_mode);
+        self
+    }
+
+    pub fn with_wal(mut self) -> Self {
+        self.is_wal = true;
+        self
+    }
+}
+
+/// Represents metadata associated with a file.
+///
+/// The `Metadata` struct defines a container for storing key-value pairs of metadata. This metadata
+/// can be used to hold additional information about a file. The data is stored in a hash map where
+/// each key is a string and each value is a vector of bytes.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Metadata {
+    /// The map holding key-value pairs of metadata.
+    ///
+    /// The `data` field is a hash map that allows associating arbitrary data with descriptive keys.
+    /// This can be used, for example, to store additional information about the file as a header.
+    data: HashMap<String, Vec<u8>>,
+}
+
+impl Metadata {
+    // Constructor for Metadata, reads data from bytes if provided
+    pub(crate) fn new(b: Option<Vec<u8>>) -> Self {
+        let mut metadata = Metadata {
+            data: HashMap::new(),
+        };
+
+        // If bytes are provided, read data from them using read_from
+        if let Some(b) = b {
+            metadata.read_from(&mut &b[..]).unwrap();
+        }
+
+        metadata
+    }
+
+    /// Creates a new `Metadata` instance with file header information.
+    ///
+    /// This method creates a new `Metadata` instance and sets the key-value pairs corresponding to
+    /// the file header information, such as magic number, version, segment ID, compression format,
+    /// and compression level.
+    ///
+    /// # Parameters
+    ///
+    /// - `id`: The segment ID value.
+    /// - `cf`: The compression format value.
+    /// - `cl`: The compression level value.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Metadata` instance containing the file header information.
+    fn new_file_header(id: u64, opts: &Options) -> Self {
+        let mut buf = Metadata::new(None);
+
+        // Set file header key-value pairs using constants
+        let cf = opts
+            .compression_format
+            .as_ref()
+            .unwrap_or(&CompressionFormat::NoCompression);
+        let cl = opts
+            .compression_level
+            .as_ref()
+            .unwrap_or(&CompressionLevel::BestSpeed);
+
+        buf.put_uint(KEY_MAGIC, MAGIC);
+        buf.put_uint(KEY_VERSION, VERSION);
+        buf.put_uint(KEY_SEGMENT_ID, id);
+        buf.put_uint(KEY_COMPRESSION_FORMAT, cf.as_u64());
+        buf.put_uint(KEY_COMPRESSION_LEVEL, cl.as_u64());
+        buf.put_uint(KEY_MAX_FILE_SIZE, opts.max_file_size);
+        if let Some(md) = opts.metadata.as_ref() {
+            buf.put(KEY_ADDITIONAL_METADATA, &md.bytes());
+        }
+
+        buf
+    }
+
+    // Returns the serialized bytes representation of Metadata
+    pub(crate) fn bytes(&self) -> Vec<u8> {
```

Comment:
I know this is not part of public API but this should be named `to_bytes` as per the naming guidelines mentioned earlier.